exchange/bitswap/wantmanager.go

```diff
@@ -0,0 +1,262 @@
+package bitswap
+
+import (
+	"sync"
+	"time"
+
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+	engine "github.com/ipfs/go-ipfs/exchange/bitswap/decision"
+	bsmsg "github.com/ipfs/go-ipfs/exchange/bitswap/message"
+	bsnet "github.com/ipfs/go-ipfs/exchange/bitswap/network"
+	wantlist "github.com/ipfs/go-ipfs/exchange/bitswap/wantlist"
+	peer "github.com/ipfs/go-ipfs/p2p/peer"
+	u "github.com/ipfs/go-ipfs/util"
+)
+
+type WantManager struct {
```

Comment:
> All non-channel member variables of the wantmanager are only touched within the context of the synchronized Run loop.

no, several members are touched by the (running in other goroutines) `runQueue`:

- https://github.com/ipfs/go-ipfs/pull/1218/files#diff-bd9ac0ba6daa6109cc6e75eda3411355R146
- https://github.com/ipfs/go-ipfs/pull/1218/files#diff-bd9ac0ba6daa6109cc6e75eda3411355R164

These are "constant" (i.e. set at construction and not changed). BUT, that's today. this is not trivial for _people who are not whyrusleeping right now_ to edit.

This, also might help:

```go
type WantManager struct {

 // channels are safe to send to by other goroutines
 incoming   chan []*bsmsg.Entry
 connect    chan peer.ID // notification for new peers connecting
 disconnect chan peer.ID // notification for peers disconnecting

 // internal members, access synchronized by run loop.
 // DO NOT access or modify outside of run loop.
 wl       *wantlist.Wantlist
 peers    map[peer.ID]*msgQueue
 receiver bsnet.Receiver
 network  bsnet.BitSwapNetwork
 ctx      context.Context
}
```